File: android/src/main/java/co/baseplay/capacitor/printer/Printer.java
package co.baseplay.capacitor.printer;

import android.content.Context;
import android.print.PrintAttributes;
import android.print.PrintDocumentAdapter;
import android.print.PrintManager;
import android.webkit.WebView;

/**
 * Print logic, kept independent of the Capacitor bridge so it can be unit
 * tested and reused from other native code.
 */
public class Printer {

    /** Job name used when the caller passes none (matches the JS default). */
    public static final String DEFAULT_JOB_NAME = "Document";

    /**
     * The job name shown in the print UI: the trimmed input, or
     * {@link #DEFAULT_JOB_NAME} when it is missing or blank.
     */
    public String resolveJobName(String name) {
        if (name == null) {
            return DEFAULT_JOB_NAME;
        }
        String trimmed = name.trim();
        return trimmed.isEmpty() ? DEFAULT_JOB_NAME : trimmed;
    }

    /**
     * Hands the WebView's current content to the Android print framework.
     *
     * The WebView's print document adapter applies the page's {@code @media print}
     * stylesheet, and the system print dialog includes the built-in "Save as PDF"
     * printer. Must be called on the UI thread.
     *
     * @param context an Activity context — PrintManager needs one to show its UI
     * @throws IllegalStateException when the device has no print service
     */
    public void printWebView(Context context, WebView webView, String jobName) {
        PrintManager printManager = (PrintManager) context.getSystemService(Context.PRINT_SERVICE);
        if (printManager == null) {
            throw new IllegalStateException("Printing is not available on this device");
        }

        PrintDocumentAdapter adapter = webView.createPrintDocumentAdapter(jobName);
        printManager.print(jobName, adapter, new PrintAttributes.Builder().build());
    }
}

File: ios/Sources/PrinterPlugin/Printer.swift
import Foundation
import UIKit
import WebKit

/// Print logic, kept independent of the Capacitor bridge so it can be unit
/// tested and reused from other native code.
@objc public class Printer: NSObject {
    /// Job name used when the caller passes none (matches the JS default).
    public static let defaultJobName = "Document"

    /// The job name shown in the print centre: the trimmed input, or
    /// `defaultJobName` when it is missing or blank.
    public func resolveJobName(_ name: String?) -> String {
        let trimmed = name?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        return trimmed.isEmpty ? Printer.defaultJobName : trimmed
    }

    /// Presents the system print sheet for the WebView's current content.
    ///
    /// The WebView's own print formatter is used, so the page's `@media print`
    /// stylesheet decides what is printed. The sheet offers AirPrint and PDF
    /// export. Completes with `true` when the user printed and `false` when
    /// they cancelled. Must be called on the main thread.
    ///
    /// - Parameter anchorView: view the sheet is anchored to on iPad, where it
    ///   is presented as a popover and needs a source rect.
    public func printWebView(
        _ webView: WKWebView,
        anchoredTo anchorView: UIView?,
        jobName: String,
        completion: @escaping (Result<Bool, Error>) -> Void
    ) {
        let printInfo = UIPrintInfo(dictionary: nil)
        printInfo.jobName = jobName
        printInfo.outputType = .general

        let controller = UIPrintInteractionController.shared
        controller.printInfo = printInfo
        controller.showsNumberOfCopies = true
        controller.printFormatter = webView.viewPrintFormatter()

        let handler: UIPrintInteractionController.CompletionHandler = { _, completed, error in
            if let error = error {
                completion(.failure(error))
            } else {
                completion(.success(completed))
            }
        }

        if UIDevice.current.userInterfaceIdiom == .pad, let anchorView = anchorView {
            let anchor = CGRect(x: anchorView.bounds.midX, y: anchorView.bounds.midY, width: 1, height: 1)
            controller.present(from: anchor, in: anchorView, animated: true, completionHandler: handler)
        } else {
            controller.present(animated: true, completionHandler: handler)
        }
    }
}
